internal/repository/master_index.go

```diff
@@ -237,24 +234,35 @@ func (mi *MasterIndex) RebuildIndex(packBlacklist restic.IDSet) (newIndex *Index
 			newIndex.Store(pb)
 		}
 
-		if !idx.Final() {
-			debug.Log("index %d isn't final, don't add to supersedes field", i)
-			continue
-		}
+		if idx.Final() {
+			id, err := idx.ID()
+			if err != nil {
+				debug.Log("index %d does not have an ID: %v", err)
+				return nil, err
+			}
 
-		id, err := idx.ID()
-		if err != nil {
-			debug.Log("index %d does not have an ID: %v", err)
-			return nil, nil, err
-		}
+			debug.Log("adding index id %v to supersedes field", id)
 
-		debug.Log("adding index id %v to supersedes field", id)
+			err = newIndex.AddToSupersedes(id)
+			if err != nil {
+				return nil, err
+			}
+			obsolete.Insert(id)
+		} else {
+			debug.Log("index %d isn't final, don't add to supersedes field", i)
+		}
 
-		err = newIndex.AddToSupersedes(id)
-		if err != nil {
-			return nil, nil, err
+		if IndexFull(newIndex) {
+			_, err := SaveIndex(ctx, repo, newIndex)
+			if err != nil {
+				return nil, err
+			}
+			newIndex = NewIndex()
 		}
-		obsolete.Insert(id)
+	}
+	_, err = SaveIndex(ctx, repo, newIndex)
```

Comment:
I think it would be a good idea to replace the index list in the master index with the rebuilt index files.

Reply:
Yes, this is a good idea! ATM `RebuildIndex` is only called from `prune` and after that call only files are deleted. So there should be no harm by this "old" master index.
I changed this and when thinking about "obsolete" indexes, I realized that there was another shortcoming: If new index files are written, the "superseded" index files are nevertheless read by `LoadIndex` (in old and new `prune` implementation). This is now also fixed and allows to first remove "superseded" index files during the prune process.